File: app/src/main/java/com/appchamp/wordchunks/data/PacksRealmHelper.java
package com.appchamp.wordchunks.data;

import com.appchamp.wordchunks.models.pojo.PackJson;
import com.appchamp.wordchunks.models.realm.Pack;
import com.appchamp.wordchunks.util.RealmUtils;

import java.util.List;

import io.realm.Realm;
import io.realm.RealmList;

import static com.appchamp.wordchunks.util.Constants.REALM_FIELD_ID;
import static com.appchamp.wordchunks.util.Constants.REALM_FIELD_STATE;

/**
 * Packs
 */
public final class PacksRealmHelper {

    private PacksRealmHelper() {
        throw new UnsupportedOperationException("u can't instantiate me...");
    }

    /**
     * Creates the list of Realm Pack objects using the PackJson list.
     *
     * @param realm
     * @param packJsonList the list of PackJson objects.
     */
    public static void createPacks(Realm realm, final List<PackJson> packJsonList) {
        RealmList<Pack> packs = new RealmList<>();
        for (int i = 0; i < packJsonList.size(); i++) {
            Pack pack = realm.createObject(Pack.class, RealmUtils.getUUID());
            PackJson packPojo = packJsonList.get(i);
            pack.setTitle(packPojo.getTitle());
            LevelsRealmHelper.createLevels(realm, pack, packPojo.getLevels());
            packs.add(pack);
        }
    }

    public static List<Pack> findAllPacks(Realm realm) {
        return realm
                .where(Pack.class)
                .findAll();
    }

    public static Pack findFirstPackByState(Realm realm, int state) {
        return realm
                .where(Pack.class)
                .equalTo(REALM_FIELD_STATE, state)
                .findFirst();
    }

    public static Pack findFirstPackById(Realm realm, String id) {
        return realm
                .where(Pack.class)
                .equalTo(REALM_FIELD_ID, id)
                .findFirst();
    }
}

File: app/src/main/java/com/appchamp/wordchunks/data/PacksRealmHelper.kt
package com.appchamp.wordchunks.data

import com.appchamp.wordchunks.models.pojo.PackJson
import com.appchamp.wordchunks.models.realm.Pack
import com.appchamp.wordchunks.util.Constants.REALM_FIELD_ID
import com.appchamp.wordchunks.util.Constants.REALM_FIELD_STATE
import io.realm.Realm
import io.realm.RealmList
import java.util.*


/**
 * Packs
 */
object PacksRealmHelper {

    /**
     * Creates the list of Realm Pack objects using the PackJson list.
     *
     * @param realm
     * @param packJsonList the list of PackJson objects.
     */
    fun createPacks(realm: Realm, packJsonList: List<PackJson>) {
        val packs = RealmList<Pack>()
        for (i in packJsonList.indices) {
            val pack = realm.createObject(Pack::class.java, UUID.randomUUID().toString())
            val packPojo = packJsonList[i]
            pack.title = packPojo.title
            pack.color = packPojo.color
            LevelsRealmHelper.createLevels(realm, pack, packPojo.levels)
            packs.add(pack)
        }
    }

    fun findAllPacks(realm: Realm): List<Pack> = realm
            .where(Pack::class.java)
            .findAll()

    fun findFirstPackByState(realm: Realm, state: Int): Pack = realm
            .where(Pack::class.java)
            .equalTo(REALM_FIELD_STATE, state)
            .findFirst()

    fun findFirstPackById(realm: Realm, id: String?): Pack = realm
            .where(Pack::class.java)
            .equalTo(REALM_FIELD_ID, id)
            .findFirst()
}
